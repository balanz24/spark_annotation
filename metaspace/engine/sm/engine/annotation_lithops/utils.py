from __future__ import annotations

import json
import logging
from base64 import urlsafe_b64encode
from hashlib import blake2b

logger = logging.getLogger('annotation-pipeline')


def jsonhash(obj) -> str:
    """
    Calculates a hash for a JSON-stringifiable object. Intended for compacting large sets of
    parameters into a simple key that can be used to distinctly identify a cache entry.

    The output is collision-resistant, but shouldn't be assumed to be cryptographically secure.
    In most cases a motivated adversary could figure out the original object contents easily, as
    there's no hidden key and it's unlikely there will be much variety in the objects hashed.
    """
    json_val = json.dumps(obj, sort_keys=True)
    hash_val = blake2b(json_val.encode(), digest_size=12).digest()
    return str(urlsafe_b64encode(hash_val), 'utf-8')
