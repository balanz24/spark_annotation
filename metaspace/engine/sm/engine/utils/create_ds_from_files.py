import json
from datetime import datetime
from pathlib import Path

from sm.engine import molecular_db
from sm.engine.dataset import Dataset


def create_ds_from_files(ds_id, ds_name, ds_input_path, config_path=None, meta_path=None):
    config_path = config_path or Path(ds_input_path) / 'config.json'
    ds_config = json.load(open(config_path))
    if 'database_ids' not in ds_config:
        ds_config['database_ids'] = [
            molecular_db.find_by_name(db, True).id for db in ds_config['databases']
        ]

    meta_path = meta_path or Path(ds_input_path) / 'meta.json'
    if not Path(meta_path).exists():
        raise Exception('meta.json not found')
    metadata = json.load(open(str(meta_path)))

    return Dataset(
        id=ds_id,
        name=ds_name,
        input_path=str(ds_input_path),
        upload_dt=datetime.now(),
        metadata=metadata,
        is_public=True,
        config=ds_config,
    )
