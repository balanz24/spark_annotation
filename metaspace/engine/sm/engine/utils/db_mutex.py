from contextlib import contextmanager
from hashlib import sha256

from psycopg2 import connect

from sm.engine.config import SMConfig

LOCK_KEY = 894951


class DBMutex:
    """
    Database-backed mutex for preventing multiple threads/processes from accessing the same
    shared resource simultaneously. Resources are specified as strings, which are simply hashed
    into a 32 bit value and locked in Postgres.
    """

    def __init__(self, dbconfig=None):
        self._dbconfig = dbconfig or SMConfig.get_conf()['db']

    @contextmanager
    def lock(self, resource_name, timeout=60):
        hash_bytes = sha256(resource_name.encode('utf-8')).digest()
        # This must be 32-bit and signed. Postgres errors if it's not a valid 32-bit signed integer
        resource_hash = int.from_bytes(hash_bytes[:4], byteorder='big', signed=True)

        conn = connect(**self._dbconfig)
        try:
            with conn.cursor() as curs:
                curs.execute('SET statement_timeout = %s;', (timeout * 1000,))
                curs.execute('SELECT pg_advisory_lock(%s, %s);', (LOCK_KEY, resource_hash))
                yield None
                curs.execute('SELECT pg_advisory_unlock(%s, %s);', (LOCK_KEY, resource_hash))
        finally:
            conn.close()
