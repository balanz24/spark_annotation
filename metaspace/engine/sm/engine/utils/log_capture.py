from logging import Handler, NOTSET, LogRecord
from contextvars import ContextVar
from contextlib import contextmanager
from typing import Optional, List, Callable

log_listeners: ContextVar[Optional[List[Callable[[str], None]]]] = ContextVar(
    'log_listeners', default=None
)


class ContextLogHandler(Handler):
    def __init__(self, level=NOTSET):
        Handler.__init__(self, level)

    def emit(self, record: LogRecord):
        listeners = log_listeners.get()
        if listeners:
            try:
                message = self.format(record)
                for listener in listeners:
                    listener(message)
            except Exception:
                self.handleError(record)


@contextmanager
def capture_logs():
    listeners = log_listeners.get()
    if not listeners:
        listeners = []
        log_listeners.set(listeners)

    captured_logs = []
    listeners.append(captured_logs.append)

    try:
        yield captured_logs
    finally:
        listeners.remove(captured_logs.append)
