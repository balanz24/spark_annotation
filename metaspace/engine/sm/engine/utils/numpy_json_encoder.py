import json
import numpy as np


class NumpyEncoder(json.JSONEncoder):
    def default(self, o):
        if isinstance(o, np.number):
            return o.item()
        elif isinstance(o, np.ndarray):
            return o.tolist()

        return json.JSONEncoder.default(self, o)


def numpy_json_dumps(obj):
    """JSON encoder that supports Numpy scalars and arrays"""
    return json.dumps(obj, cls=NumpyEncoder)
