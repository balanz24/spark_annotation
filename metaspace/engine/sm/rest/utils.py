import json

import bottle

OK = {'status_code': 200, 'status': 'success'}
NOT_EXIST = {'status_code': 404, 'status': 'not_exist'}
ALREADY_EXISTS = {'status_code': 400, 'status': 'already_exists'}
WRONG_PARAMETERS = {'status_code': 400, 'status': 'wrong_parameters'}
INTERNAL_ERROR = {'status_code': 500, 'status': 'server_error'}


def body_to_json(request):
    body = request.body.getvalue()
    return json.loads(body.decode('utf-8'))


def make_response(status_doc, **kwargs):
    bottle.response.status = status_doc['status_code']
    return {'status': status_doc['status'], **kwargs}
