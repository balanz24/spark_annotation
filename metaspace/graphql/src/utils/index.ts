export { createConnection, findUserByEmail } from './db'
export { smApiDatasetRequest } from './smApi/datasets'

export type LooselyCompatible<T> =
  {[K in keyof T]?: T[K] extends (string | number | null | undefined) ? T[K] | null : any};
