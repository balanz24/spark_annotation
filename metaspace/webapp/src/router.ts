/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter, { RawLocation } from 'vue-router'
import AboutPage from './modules/App/AboutPage'
import DatasetsPage from './modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from './modules/Account'
import { redirectAfterSignIn } from './modules/Account/signInReturnUrl'
import { updateDBParam } from './modules/Filters/url'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Annotations/AnnotationsPage.vue'),
  DatasetSummary: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ './modules/Datasets/summary/DatasetSummary.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ './modules/MetadataEditor/MetadataEditPage.vue'),
  ImageAlignmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ './modules/ImageAlignment/ImageAlignmentPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ './modules/MetadataEditor/UploadPage.vue'),
  DatasetOverviewPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/overview/DatasetOverviewPage'),
  DatasetComparisonPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/comparison/DatasetComparisonPage'),
  DatasetBrowserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/imzml/DatasetBrowserPage'),
  DatasetEnrichmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/enrichment/DatasetEnrichmentPage'),
  SpottingProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "SpottingProjectPage" */ './modules/SpottingProject/DashboardPage'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/HelpPage.vue'),
  EditUserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Group/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Group/GroupsListPage'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/TermsPage.vue'),
  PublicationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PublicationsPage'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Group/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Project/ViewProjectPage.vue'),

  // Separate bundle for design docs
  DesignTOC: () => import(/* webpackChunkName: "DesignBundle" */ './design/TOCPage.vue'),
  DesignStyleGuide: () => import(/* webpackChunkName: "DesignBundle" */ './design/StyleGuidePage.vue'),
  DesignIcons: () => import(/* webpackChunkName: "DesignBundle" */ './design/IconsPage.vue'),
  DesignComponents: () => import(/* webpackChunkName: "DesignBundle" */ './design/ComponentsPage.vue'),
  DesignForms: () => import(/* webpackChunkName: "DesignBundle" */ './design/FormsPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(about)?', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
    { path: '/annotations', name: 'annotations', component: asyncPages.AnnotationsPage },
    {
      path: '/datasets',
      component: DatasetsPage,
      children: [
        { path: '', component: asyncPages.DatasetTable },
        { path: 'summary', component: asyncPages.DatasetSummary },
      ],
    },
    {
      path: '/datasets/:dataset_id/comparison',
      name: 'datasets-comparison',
      component: asyncPages.DatasetComparisonPage,
    },
    { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
    { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.ImageAlignmentPage },
    { path: '/dataset/:dataset_id', name: 'dataset-overview', component: asyncPages.DatasetOverviewPage },
    { path: '/dataset/:dataset_id/annotations', name: 'dataset-annotations', component: asyncPages.AnnotationsPage },
    { path: '/dataset/:dataset_id/browser', name: 'dataset-browser', component: asyncPages.DatasetBrowserPage },
    { path: '/dataset/:dataset_id/enrichment', name: 'dataset-enrichment', component: asyncPages.DatasetEnrichmentPage },
    { path: '/upload', component: asyncPages.UploadPage },
    { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },
    { path: '/user/me', component: asyncPages.EditUserPage },

    { path: '/admin/health', component: asyncPages.SystemHealthPage },

    { path: '/account/sign-in', component: DialogPage, props: { dialog: 'signIn' } },
    { path: '/account/sign-in-success', redirect: redirectAfterSignIn },
    { path: '/account/create-account', component: DialogPage, props: { dialog: 'createAccount' } },
    { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
    { path: '/account/reset-password', component: ResetPasswordPage },

    { path: '/groups', component: asyncPages.GroupsListPage },
    { path: '/group/create', component: asyncPages.CreateGroupPage },
    { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

    { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
    { // Legacy URL sent in "request access" emails up until Feb 2019
      path: '/project/:projectIdOrSlug/manage',
      redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } } as RawLocation,
    },
    { path: '/projects', component: asyncPages.ProjectsListPage },

    { path: '/terms', component: asyncPages.TermsPage, meta: { footer: true } },
    { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
    { path: '/publications', component: asyncPages.PublicationsPage, meta: { footer: true } },

    { path: '/design', component: asyncPages.DesignTOC, meta: { footer: true, flex: true } },
    { path: '/design/styleguide', component: asyncPages.DesignStyleGuide, meta: { footer: true, flex: true } },
    { path: '/design/icons', component: asyncPages.DesignIcons, meta: { footer: true, flex: true } },
    { path: '/design/components', component: asyncPages.DesignComponents, meta: { footer: true, flex: true } },
    { path: '/design/forms', component: asyncPages.DesignForms, meta: { footer: true, flex: true } },

    { path: '/detectability', name: 'detectability', component: asyncPages.SpottingProjectPage },

    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const pageLoadedAt = Date.now()
router.beforeEach((to, from, next) => {
  // If user has had the window open for a long time, it's likely they're using old code. This can cause issues
  // such as incorrect API usage, or causing error reports for bugs that have already been fixed.
  // Force the browser to reload the page, preferring to do so during page transitions to minimize impact.
  const daysSincePageLoad = (Date.now() - pageLoadedAt) / 1000 / 60 / 60 / 24
  if (to.path !== from.path && daysSincePageLoad > 1.75 || daysSincePageLoad > 7) {
    window.location.assign(to.fullPath)
    next(false)
  } else {
    next()
  }
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
