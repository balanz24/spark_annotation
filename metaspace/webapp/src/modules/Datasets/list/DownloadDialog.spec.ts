import { mount } from '@vue/test-utils'
import router from '../../../router'
import { initMockGraphqlClient, apolloProvider } from '../../../../tests/utils/mockGraphqlClient'
import Vue from 'vue'
import DownloadDialog from './DownloadDialog'

const TestDownloadDialog = Vue.component('download-dialog-wrapper', {
  functional: true,
  render: (h, { props }) => h(DownloadDialog, { props }),
})

describe('Datasets/DownloadDialog', () => {
  it('should match snapshot', async() => {
    initMockGraphqlClient({
      Query: () => ({
        dataset: () => {
          return {
            downloadLinkJson: JSON.stringify({
              contributors: [
                { name: 'Foo Bar', institution: 'EMBL' },
              ],
              license: {
                code: 'CC BY 4.0',
                name: 'Creative Commons Attribution 4.0 International Public License',
                link: 'https://creativecommons.org/licenses/by/4.0/',
              },
              files: [
                {
                  filename: 'Image5.ibd',
                  link: 'https://sm-engine-upload.s3.eu-west-1.amazonaws.com/test/Image5.ibd?AWSAccessKeyId=test'
                    + '&Expires=1585569522&Signature=test%3D',
                },
                {
                  filename: 'Image5.imzML',
                  link: 'https://sm-engine-upload.s3.eu-west-1.amazonaws.com/test/Image5.imzML?AWSAccessKeyId=test'
                    + '&Expires=1585569522&Signature=test%3D',
                },
              ],
            }),
          }
        },
      }),
    })
    const propsData = {
      datasetId: '2020-01-02_03h04m05s',
      datasetName: 'Mouse whole body DHB',
    }
    const wrapper = mount(TestDownloadDialog, { router, apolloProvider, propsData })
    await Vue.nextTick()

    expect(wrapper.element).toMatchSnapshot()
  })
})
