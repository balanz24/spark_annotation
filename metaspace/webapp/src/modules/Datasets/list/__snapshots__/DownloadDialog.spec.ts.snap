// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datasets/DownloadDialog should match snapshot 1`] = `
<transition-stub
  name="dialog-fade"
  style="z-index: 2001;"
>
  <div
    class="el-dialog__wrapper"
  >
    <div
      aria-label="Download Mouse whole body DHB"
      aria-modal="true"
      class="el-dialog el-dialog-lean max-w-xl"
      role="dialog"
      style="margin-top: 15vh;"
    >
      <div
        class="el-dialog__header"
      >
        <span
          class="el-dialog__title"
        >
          Download Mouse whole body DHB
        </span>
        <button
          aria-label="Close"
          class="el-dialog__headerbtn"
          type="button"
        >
          <i
            class="el-dialog__close el-icon el-icon-close"
          />
        </button>
      </div>
      <div
        class="el-dialog__body"
      >
        <p>
          <p>
            This dataset is distributed under the 
            <a
              href="https://creativecommons.org/licenses/by/4.0/"
              rel="noopener noreferrer nofollow"
              target="_blank"
            >
              CC BY 4.0
            </a>
             license. Make sure to credit the author of this dataset:
          </p>
        </p>
        <ul>
          <li
            class="list-none"
          >
            Foo Bar, EMBL
          </li>
        </ul>
        <h4>
          Files
        </h4>
        <div
          class="flex my-1 flex-wrap"
        >
          <div
            class="flex items-center p-1 w-1/2 box-border"
          >
            <a
              href="https://sm-engine-upload.s3.eu-west-1.amazonaws.com/test/Image5.ibd?AWSAccessKeyId=test&Expires=1585569522&Signature=test%3D"
            >
              <svg
                style="width: 32px; height: 50px;"
                viewBox="0 0 72 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="a"
                  x1="36"
                  x2="36"
                  y1="99"
                  y2="1"
                >
                  <stop
                    offset="0"
                    stop-color="#c8d4db"
                  />
                  <stop
                    offset=".139"
                    stop-color="#d8e1e6"
                  />
                  <stop
                    offset=".359"
                    stop-color="#ebf0f3"
                  />
                  <stop
                    offset=".617"
                    stop-color="#f9fafb"
                  />
                  <stop
                    offset="1"
                    stop-color="#fff"
                  />
                </lineargradient>
                <path
                  d="M45 1L71 27.7V99H1V1H45Z"
                  fill="url(#a)"
                  stroke="#7191a1"
                  stroke-width="2"
                />
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="b"
                  x1="45.068"
                  x2="58.568"
                  y1="27.796"
                  y2="14.395"
                >
                  <stop
                    offset="0"
                    stop-color="#fff"
                  />
                  <stop
                    offset=".35"
                    stop-color="#fafbfb"
                  />
                  <stop
                    offset=".532"
                    stop-color="#edf1f4"
                  />
                  <stop
                    offset=".675"
                    stop-color="#dde5e9"
                  />
                  <stop
                    offset=".799"
                    stop-color="#c7d3da"
                  />
                  <stop
                    offset=".908"
                    stop-color="#adbdc7"
                  />
                  <stop
                    offset="1"
                    stop-color="#92a5b0"
                  />
                </lineargradient>
                <path
                  d="M45 1L71 27.7H45V1z"
                  fill="url(#b)"
                  stroke="#7191a1"
                  stroke-linejoin="bevel"
                  stroke-width="2"
                />
                <text
                  fill="#7191a1"
                  font-size="20px"
                  font-weight="bold"
                  text-anchor="middle"
                  text-length="64"
                  x="36"
                  y="60"
                >
                  .ibd
                </text>
              </svg>
            </a>
            <a
              class="flex mx-3 my-1 overflow-hidden"
              href="https://sm-engine-upload.s3.eu-west-1.amazonaws.com/test/Image5.ibd?AWSAccessKeyId=test&Expires=1585569522&Signature=test%3D"
            >
              <span
                class="flex-shrink truncate"
              >
                Image5
              </span>
              <span>
                .ibd
              </span>
            </a>
          </div>
          <div
            class="flex items-center p-1 w-1/2 box-border"
          >
            <a
              href="https://sm-engine-upload.s3.eu-west-1.amazonaws.com/test/Image5.imzML?AWSAccessKeyId=test&Expires=1585569522&Signature=test%3D"
            >
              <svg
                style="width: 32px; height: 50px;"
                viewBox="0 0 72 100"
                xmlns="http://www.w3.org/2000/svg"
              >
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="a"
                  x1="36"
                  x2="36"
                  y1="99"
                  y2="1"
                >
                  <stop
                    offset="0"
                    stop-color="#c8d4db"
                  />
                  <stop
                    offset=".139"
                    stop-color="#d8e1e6"
                  />
                  <stop
                    offset=".359"
                    stop-color="#ebf0f3"
                  />
                  <stop
                    offset=".617"
                    stop-color="#f9fafb"
                  />
                  <stop
                    offset="1"
                    stop-color="#fff"
                  />
                </lineargradient>
                <path
                  d="M45 1L71 27.7V99H1V1H45Z"
                  fill="url(#a)"
                  stroke="#7191a1"
                  stroke-width="2"
                />
                <lineargradient
                  gradientUnits="userSpaceOnUse"
                  id="b"
                  x1="45.068"
                  x2="58.568"
                  y1="27.796"
                  y2="14.395"
                >
                  <stop
                    offset="0"
                    stop-color="#fff"
                  />
                  <stop
                    offset=".35"
                    stop-color="#fafbfb"
                  />
                  <stop
                    offset=".532"
                    stop-color="#edf1f4"
                  />
                  <stop
                    offset=".675"
                    stop-color="#dde5e9"
                  />
                  <stop
                    offset=".799"
                    stop-color="#c7d3da"
                  />
                  <stop
                    offset=".908"
                    stop-color="#adbdc7"
                  />
                  <stop
                    offset="1"
                    stop-color="#92a5b0"
                  />
                </lineargradient>
                <path
                  d="M45 1L71 27.7H45V1z"
                  fill="url(#b)"
                  stroke="#7191a1"
                  stroke-linejoin="bevel"
                  stroke-width="2"
                />
                <text
                  fill="#7191a1"
                  font-size="20px"
                  font-weight="bold"
                  text-anchor="middle"
                  text-length="64"
                  x="36"
                  y="60"
                >
                  .imzML
                </text>
              </svg>
            </a>
            <a
              class="flex mx-3 my-1 overflow-hidden"
              href="https://sm-engine-upload.s3.eu-west-1.amazonaws.com/test/Image5.imzML?AWSAccessKeyId=test&Expires=1585569522&Signature=test%3D"
            >
              <span
                class="flex-shrink truncate"
              >
                Image5
              </span>
              <span>
                .imzML
              </span>
            </a>
          </div>
        </div>
        <p>
          <i>
            These download links expire after 30 minutes.
          </i>
        </p>
      </div>
      <!---->
    </div>
  </div>
</transition-stub>
`;
